### PR TITLE
producer: make Flush.Frequency behaviour better

### PR DESCRIPTION
**Pull request by eapache**
Previously the timer would tick every Flush.Frequency regardless of if a message
was actually queued. This meant that if no message arrived for Flush.Frequency
then the next message would be sent immediately, which is unexpected if not
*exactly* wrong.

With this change the timer is only started when the first message arrives, and
is cleared when a flush occurs. This should result in slightly better batching
for low-volume topics at the result of slightly higher latency (although the
delay will still never be more than Flush.Frequency).

@Shopify/kafka

### CODE REVIEW

File: async_producer.go [hunk 1]
```diff
@@ -421,17 +421,13 @@ func (p *asyncProducer) leaderDispatcher(topic string, partition int32, input ch
...
-	}
-
-	var buffer []*ProducerMessage
-	var doFlush chan []*ProducerMessage
-	var bytesAccumulated int
-	var defaultFlush bool
+	var (
+		timer            <-chan time.Time
+		buffer           []*ProducerMessage
+		doFlush          chan []*ProducerMessage
```
Comment (wvanbergen) [inline async_producer.go]: `flushQueue` ?

**Comment by wvanbergen** [comment]
Looks innocent enough to me :)

File: async_producer.go (hunk 1 above)
Comment (eapache) [inline async_producer.go]: `flushTriggered`?

Comment (wvanbergen) [inline async_producer.go]: `triggerFush`! :P

**Pull request merged**